### ISSUE TITLE
Standard for documentation types

### ISSUE THREAD

**Issue by ColCarroll** [arviz-devs/arviz#415]
## Tell us about it

Documentation uses `numpydoc` and is linted using `pydocstyle`. However, the format of the type is not checked.  Specifically, [densityplot.py](https://github.com/arviz-devs/arviz/blob/212c5977aba29be8f0ed25cf21e05ffac916dc7b/arviz/plots/densityplot.py#L40) now uses proper type hinting, while everywhere else uses a less rigorous style emulating other pydata projects (`int, optional`, instead of `Optional[int]`, for example)

I am most in favor of consistency across the code base, which would suggest changing densityplot.py. I would also prefer type hinting everywhere to our current standard.

**Specifically**: I think we should manually urge new contributions to match current documentation practices, but I would also accept a pull request that converts a reasonable portion of the codebase to the new style.

## Thoughts on implementation

My feelings on this are weak enough that I *do not* think it should be in CONTRIBUTING.md, since every requirement there feels like a hurdle to engagement. If we ended up in a place with type hinting everywhere, then I would be in favor of adding it.

**Comment by canyon289** [comment]
A couple points
1. Putting something that looks like type hints in the docstring isn't actually type hints. Type hints (mostly) go in the argument list for a function. From everything I can read numpy docstrings and type hinting can coexist and don't clobber each other. I have a weak prior on this though and am continuing to read

2. I'm strongly in favor of adopting keeping numpy style docstrings. Some of our major dependencies  call for them explicitly and I don't see a good reason to diverge from our larger community

https://matplotlib.org/devel/documenting_mpl.html#formatting-conventions
http://xarray.pydata.org/en/stable/contributing.html#about-the-xarray-documentation
https://docs.python.org/3/library/typing.html

3. I do agree that docstring style does pose a barrier. Perhaps as a compromise we can accept pull requests where the docstring style doesn't match but make an issue to go back and fix it

**Comment by OriolAbril** [comment]
I vote for numpy style docstrings and leaving the issue open to encourage anybody who might feel like it to use type hinting in the function arguments. It would have some advantages, mainly when using IDEs as far as I know.

**Comment by OriolAbril** [comment]
Should probably be added to #798 so it is clear what should be followed.

**Comment by percygautam** [comment]
@OriolAbril So, we have to extend the contribution guide about new links suggested by @canyon289 . This can also be added to the new PR template being added. Also, are you suggesting to change all docstrings to `mypy` format?

**Comment by OriolAbril** [comment]
`mypy` is to be used in the code itself, not in the docstrings. See example below.

```
def from_pymc3(
    trace: MultiTrace, prior: Optional[Dict[str, np.ndarray]]=None, ...
) -> InferenceData:
    """
    Converts ...

    Parameters:
    trace: pymc3.MultiTrace
        description
    prior: dict of {str: np.ndarray}, optional
        description
    ...
    """
    code
    ...
```

**Comment by percygautam** [comment]
@OriolAbril Thanks for the clarification. I'll come back to this issue in some days. I'll first like to fully explore all the data types ( like 'InferenceData') as I seem to be lacking.

**Comment by OriolAbril** [comment]
I think we can close this after #1065
